File: autoit-mediapipe-com/src/binding/tasks/text/text_classifier.cc
#include "binding/tasks/text/text_classifier.h"
#include "binding/packet_getter.h"

namespace {
	using namespace mediapipe::tasks::text::text_classifier::proto;
	using namespace mediapipe::tasks::autoit::components::containers::classification_result;
	using namespace mediapipe::tasks::autoit::core::base_options;
	using namespace mediapipe::tasks::autoit::core::task_info;
	using namespace mediapipe::autoit::packet_getter;

	const std::string _CLASSIFICATIONS_STREAM_NAME = "classifications_out";
	const std::string _CLASSIFICATIONS_TAG = "CLASSIFICATIONS";
	const std::string _TEXT_IN_STREAM_NAME = "text_in";
	const std::string _TEXT_TAG = "TEXT";
	const std::string _TASK_GRAPH_NAME = "mediapipe.tasks.text.text_classifier.TextClassifierGraph";
}

namespace mediapipe::tasks::autoit::text::text_classifier {
	absl::StatusOr<std::shared_ptr<TextClassifierGraphOptions>> TextClassifierOptions::to_pb2() const {
		auto pb2_obj = std::make_shared<TextClassifierGraphOptions>();

		if (base_options) {
			MP_ASSIGN_OR_RETURN(auto base_options_proto, base_options->to_pb2());
			pb2_obj->mutable_base_options()->CopyFrom(*base_options_proto);
		}

		if (score_threshold) pb2_obj->mutable_classifier_options()->set_score_threshold(*score_threshold);
		if (category_allowlist) pb2_obj->mutable_classifier_options()->mutable_category_allowlist()->Add(category_allowlist->begin(), category_allowlist->end());
		if (category_denylist) pb2_obj->mutable_classifier_options()->mutable_category_denylist()->Add(category_denylist->begin(), category_denylist->end());
		if (display_names_locale) pb2_obj->mutable_classifier_options()->set_display_names_locale(*display_names_locale);
		if (max_results) pb2_obj->mutable_classifier_options()->set_max_results(*max_results);

		return pb2_obj;
	}

	absl::StatusOr<std::shared_ptr<TextClassifier>> TextClassifier::create(
		const CalculatorGraphConfig& graph_config
	) {
		using BaseTextTaskApi = core::base_text_task_api::BaseTextTaskApi;
		return BaseTextTaskApi::create(graph_config, static_cast<TextClassifier*>(nullptr));
	}

	absl::StatusOr<std::shared_ptr<TextClassifier>> TextClassifier::create_from_model_path(const std::string& model_path) {
		auto base_options = std::make_shared<BaseOptions>(model_path);
		return create_from_options(std::make_shared<TextClassifierOptions>(base_options));
	}

	absl::StatusOr<std::shared_ptr<TextClassifier>> TextClassifier::create_from_options(std::shared_ptr<TextClassifierOptions> options) {
		TaskInfo task_info;
		task_info.task_graph = _TASK_GRAPH_NAME;
		*task_info.input_streams = { _TEXT_TAG + ":" + _TEXT_IN_STREAM_NAME };
		*task_info.output_streams = { _CLASSIFICATIONS_TAG + ":" + _CLASSIFICATIONS_STREAM_NAME };
		MP_ASSIGN_OR_RETURN(task_info.task_options, options->to_pb2());

		MP_ASSIGN_OR_RETURN(auto config, task_info.generate_graph_config());

		return create(*config);
	}

	absl::StatusOr<std::shared_ptr<TextClassifierResult>> TextClassifier::classify(const std::string& text) {
		MP_ASSIGN_OR_RETURN(auto output_packets, _runner->Process({
			{ _TEXT_IN_STREAM_NAME, std::move(MakePacket<std::string>(text)) }
			}));

		MP_PACKET_ASSIGN_OR_RETURN(const auto& classification_result_proto, mediapipe::tasks::components::containers::proto::ClassificationResult, output_packets.at(_CLASSIFICATIONS_STREAM_NAME));
		return TextClassifierResult::create_from_pb2(classification_result_proto);
	}
}

File: src/declarations/tasks/text/text_classifier.js
const ns = "mediapipe::tasks::text::text_classifier";
const nsProgId = ns.replaceAll("::", ".");
const fqn = `${ ns }::TextClassifier`;

module.exports = ({language, self_get}) => {
    const progid = `mediapipe.tasks.${ language }.text.text_classifier`;
    const utils = `mediapipe::tasks::${ language }::core::utils`;

    return [
        // expose text_classifier properties like in mediapipe python
        [`${ progid }.`, "", ["/Properties"], [
            [`${ ns }::TextClassifierOptions`, "TextClassifierOptions", "", ["/R", "=this", "/S"]],
            [`${ ns }::TextClassifier`, "TextClassifier", "", ["/R", "=this", "/S"]],
            ["mediapipe::tasks::components::containers::ClassificationResult", "TextClassifierResult", "", ["/R", "=this", "/S"]],
        ], "", ""],

        [`struct ${ nsProgId }.TextClassifierOptions`, "", [`/progid=${ progid }.TextClassifierOptions`, "/Simple"], [
            ["tasks::core::BaseOptions", "base_options", "", ["/RW"]],
            ["components::processors::ClassifierOptions", "classifier_options", "", ["/RW"]],

            ["std::string", "display_names_locale", "", ["/RW", `/RExpr=${ self_get("classifier_options") }.display_names_locale`, `/WExpr=autoit_to($value, ${ self_get("classifier_options") }.display_names_locale)`]],
            ["int", "max_results", "", ["/RW", `/RExpr=${ self_get("classifier_options") }.max_results`, `/WExpr=${ self_get("classifier_options") }.max_results = $value`]],
            ["float", "score_threshold", "", ["/RW", `/RExpr=${ self_get("classifier_options") }.score_threshold`, `/WExpr=${ self_get("classifier_options") }.score_threshold = $value`]],
            ["std::vector<std::string>", "category_allowlist", "", ["/RW", `/RExpr=${ self_get("classifier_options") }.category_allowlist`, `/WExpr=autoit_to($value, ${ self_get("classifier_options") }.category_allowlist)`]],
            ["std::vector<std::string>", "category_denylist", "", ["/RW", `/RExpr=${ self_get("classifier_options") }.category_denylist`, `/WExpr=autoit_to($value, ${ self_get("classifier_options") }.category_denylist)`]],
        ], "", ""],

        [`${ nsProgId }.TextClassifierOptions.TextClassifierOptions`, "", ["/Expr=", `/DC=
            if (base_options) { ${ self_get("base_options") } = std::move(*base_options); }
            if (display_names_locale) { ${ self_get("classifier_options") }.display_names_locale = std::move(*display_names_locale); }
            if (max_results) { ${ self_get("classifier_options") }.max_results = std::move(*max_results); }
            if (score_threshold) { ${ self_get("classifier_options") }.score_threshold = std::move(*score_threshold); }
            if (category_allowlist) { ${ self_get("classifier_options") }.category_allowlist = std::move(*category_allowlist); }
            if (category_denylist) { ${ self_get("classifier_options") }.category_denylist = std::move(*category_denylist); }
            if (classifier_options) { ${ self_get("classifier_options") } = std::move(*classifier_options); }
        `.replace(/^ {8}/mg, "").trim()], [
            ["std::optional<tasks::core::BaseOptions>", "base_options", "std::nullopt", []],
            ["std::optional<std::string>", "display_names_locale", "std::nullopt", []],
            ["std::optional<int>", "max_results", "std::nullopt", []],
            ["std::optional<float>", "score_threshold", "std::nullopt", []],
            ["std::optional<std::vector<std::string>>", "category_allowlist", "std::nullopt", []],
            ["std::optional<std::vector<std::string>>", "category_denylist", "std::nullopt", []],
            ["std::optional<components::processors::ClassifierOptions>", "classifier_options", "std::nullopt", []],
        ], "", ""],

        [`class ${ nsProgId }.TextClassifier`, "", [`/progid=${ progid }.TextClassifier`], [], "", ""],

        [`${ nsProgId }.TextClassifier.Create`, `std::shared_ptr<${ fqn }>`, ["=create_from_model_path", "/S", `/WrapAs=${ utils }::ReleaseToSharedPtr`], [
            ["std::string", "model_path", "", ["/Expr=std::unique_ptr<TextClassifierOptions>(new TextClassifierOptions{ .base_options{ .model_asset_path{ $0 } } })"]],
        ], "", ""],

        [`${ nsProgId }.TextClassifier.Create`, `std::shared_ptr<${ fqn }>`, ["=create_from_options", "/S", `/WrapAs=${ utils }::ReleaseToSharedPtr`], [
            ["std::shared_ptr<TextClassifierOptions>", "options", "", ["/Expr=tasks::autoit::core::utils::ConvertToUniquePtr($0)"]],
        ], "", ""],

        [`${ nsProgId }.TextClassifier.Classify`, "absl::StatusOr<TextClassifierResult>", ["=classify"], [
            ["std::string", "text", "", []],
        ], "", ""],

        [`${ nsProgId }.TextClassifier.Close`, "absl::Status", ["=close"], [], "", ""],
    ];
};
